esp-link/cgitelnet.c

```diff
@@ -0,0 +1,85 @@
+#include <esp8266.h>
+#include "cgi.h"
+#include "config.h"
+#include "serbridge.h"
+
+// Cgi to return choice of Telnet ports
+int ICACHE_FLASH_ATTR cgiTelnetGet(HttpdConnData *connData) {
+  if (connData->conn==NULL) return HTTPD_CGI_DONE; // Connection aborted
+
+  char buff[1024];
+  int len;
+  
+  os_printf("Current telnet ports: port1=%d port2=%d\n",
+	flashConfig.telnet_port1, flashConfig.telnet_port2);
+	
+  len = os_sprintf(buff,
+      "{ \"port1\": \"%d\", \"port2\": \"%d\" }",
+      flashConfig.telnet_port1, flashConfig.telnet_port2);
+
+  jsonHeader(connData, 200);
+  httpdSend(connData, buff, len);
+  return HTTPD_CGI_DONE;
+}
+
+// Cgi to change choice of Telnet ports
+int ICACHE_FLASH_ATTR cgiTelnetSet(HttpdConnData *connData) {
+  if (connData->conn==NULL) {
+    return HTTPD_CGI_DONE; // Connection aborted
+  }
+	
+  char *coll;
+  int8_t ok = 0;
+  uint16_t port1, port2;
+  ok |= getUInt16Arg(connData, "port1", &port1);
+  ok |= getUInt16Arg(connData, "port2", &port2);
+  if (ok < 0) { coll = "Failed to fetch ports. Are they valid?"; goto collision; }
+
+  if (ok > 0) {
+    // fill both port variables from flash or ajax provided value
+    if (!port1) port1 = flashConfig.telnet_port1;
+    if (!port2) port2 = flashConfig.telnet_port2;
+  
+    // check whether ports are different
+    if (port1 == port2) { coll = "Ports cannot be the same!"; goto collision; }
+  
+    // we're good, set flashconfig
+    flashConfig.telnet_port1 = port1;
+    flashConfig.telnet_port2 = port2;
+    os_printf("Telnet ports changed: port1=%d port2=%d\n",
+	flashConfig.telnet_port1, flashConfig.telnet_port2);
+
+    // save to flash
+    if (configSave()) {
+      httpdStartResponse(connData, 204);
+      httpdEndHeaders(connData);
+    } else {
+      httpdStartResponse(connData, 500);
+      httpdEndHeaders(connData);
+      httpdSend(connData, "Failed to save config", -1);
+    }
+    
+    // apply the changes
+    serbridgeInit(flashConfig.telnet_port1, flashConfig.telnet_port2);
```

Comment:
Something that is missing here is to call a new serbridgeStop function that closes the current listening sockets and resets any serbridge state so you can call serbridgeInit again. The latter wasn't designed to be called multiple times...
This may also explain some of the issues you have been seeing.